lib/process/watch.go

```diff
@@ -97,3 +99,150 @@ func (p *Process) watchCertificate(ctx context.Context, client *kubernetes.Clien
+}
+
+// startWatchingAuthGatewayEvents launches watcher that monitors auth
+// gateway configuration change events and appropriately updates
+// Teleport configuration.
+func (p *Process) startWatchingAuthGatewayEvents(ctx context.Context, client *kubernetes.Clientset) error {
+	go func() {
+		eventsCh := make(chan service.Event)
+		p.WaitForEvent(ctx, constants.AuthGatewayConfigUpdatedEvent, eventsCh)
+		p.Infof("Started watching %v events.", constants.AuthGatewayConfigUpdatedEvent)
+		for {
+			select {
+			case event := <-eventsCh:
+				if event.Name != constants.AuthGatewayConfigUpdatedEvent {
+					p.Warnf("Expected %v event, got: %#v.", constants.AuthGatewayConfigUpdatedEvent, event)
+					continue
+				}
+				p.Infof("Received event: %#v.", event)
+				authGatewayConfig, err := p.getAuthGatewayConfig()
+				if err != nil {
+					p.Errorf(trace.DebugReport(err))
+					continue
+				}
+				if authGatewayConfig.PrincipalsChanged(p.authGatewayConfig) {
+					// Teleport principals got updated. Don't restart right
+					// away, but update its config so it can regenerate
+					// identities for its services.
+					p.Infof("Auth gateway principals changed.")
+					config, err := p.buildTeleportConfig()
+					if err != nil {
+						p.Errorf(trace.DebugReport(err))
+						continue
+					}
+					// Replacing principals in config will result in Teleport
+					// regenerating identities (asynchonously) and then
+					// sending reload event which will be caught below.
+					processconfig.ReplacePublicAddrs(p.teleportProcess().Config, config)
+				} else if authGatewayConfig.SettingsChanged(p.authGatewayConfig) {
+					// Principals didn't change but some of the Teleport
+					// settings changed so we can reload right away.
+					p.Infof("Auth gateway settings changed.")
+					p.BroadcastEvent(service.Event{
+						Name: service.TeleportReloadEvent,
+					})
+				} else {
+					// Neither principals nor other settings changed, nothing
+					// to do (maybe auth preference changed which is also a
+					// part auth gateway resource).
+					p.Infof("Auth gateway principals/settings didn't change.")
+				}
+			case <-ctx.Done():
+				p.Infof("Stopped watching %v events.", constants.AuthGatewayConfigUpdatedEvent)
+				return
+			}
+		}
+	}()
+	return nil
+}
+
+// startWatchingReloadEvents launches watcher that listens for reload events
+// and restarts the process.
+func (p *Process) startWatchingReloadEvents(ctx context.Context, client *kubernetes.Clientset) error {
+	go func() {
+		eventsCh := make(chan service.Event)
+		p.WaitForEvent(ctx, service.TeleportReloadEvent, eventsCh)
+		p.Infof("Started watching %v events.", service.TeleportReloadEvent)
+		for {
+			select {
+			case event := <-eventsCh:
+				if event.Name != service.TeleportReloadEvent {
+					p.Warnf("Expected %v event, got: %#v.", service.TeleportReloadEvent, event)
+					continue
+				}
+				p.Infof("Received event: %#v.", event)
+				err := libkube.DeleteSelf(client, p.FieldLogger)
```

Comment:
Well, teleport has a hot-reload feature but I'm not sure how hard would be to integrate it into gravity (or even if possible at all with it running inside k8s and all) or if that's worth the time. I have also originally attempted to perform a graceful shutdown instead of deleting the pod, but that just leads to the pod going into crashloop as k8s does not distinguish b/w 0 and non-0 exit codes.